ops/main.py

```diff
@@ -42,6 +43,134 @@
 logger = logging.getLogger()
 
 
+@dataclasses.dataclass(frozen=True)
+class _JujuContext:
+    # source: https://juju.is/docs/juju/charm-environment-variables
+    # the HookVars function: https://github.com/juju/juju/blob/3.6/worker/uniter/runner/context/context.go#L1398
+    # Only a subset of the above source, because these are what are used in ops.
+
+    # The action's name, for example 'backup' (from JUJU_ACTION_NAME).
+    action_name: Optional[str] = None
+
+    # The action's uuid, for example '1' (from JUJU_ACTION_UUID).
+    action_uuid: Optional[str] = None
+
+    # The root directory of the charm where it is running,
+    # for example '/var/lib/juju/agents/unit-bare-0/charm' (from JUJU_CHARM_DIR).
+    # If JUJU_CHARM_DIR is None or set to an empty string,
+    # use Path(f'{__file__}/../../..') as default
+    # (assuming the '$JUJU_CHARM_DIR/lib/op/main.py' structure).
+    charm_dir: Path = Path()
```

Comment:
Then it will be the same as `JujuVersion` (see the other comment above), Ruff will complain "Do not perform function call in dataclass defaults" Ruff[RUF009](https://docs.astral.sh/ruff/rules/function-call-in-dataclass-default-argument). Shall we do it?

Reply:
You'd do it like:

```suggestion
    charm_dir: Path = dataclasses.field(default_factory=lambda: Path(f'{__file__}'/../../..'))
```

Although I haven't checked that that's exactly the right path. But using `default_factory`.